azavea/tilegarden
# Refactor error handling to dynamically decide status codes

## Overview

This PR removes the previous error handling system and replaces it with one that returns variable HTTP status codes depending on the situation. The goal was to be able to give users/tools that use the API more feedback about things going wrong, as well as intercept Postgis errors in order to prevent database info leaking out through error messages. Additionally, the removal of `jimp` as a dependency slims down the bundle size, which is a nice side effect. 

Error cases that I added:
 * Conversion of Postgis errors to simply "Postgis error" to prevent database info leakage
 * 404 response on all undefined endpoints
 * 400 status if there doesn't exist a layer for a given value of '&layers'
 * Misc error handlers for forseen common API typos

### Demo

![image](https://user-images.githubusercontent.com/16990679/43969816-618c7c48-9c99-11e8-9f4a-5278b0983bf2.png)



### Notes
 * CloudFront will often not forward along HTTP statuses from API Gateway, returning a 500 'Internal server error' status rather than the actual, more specific error that API Gateway sends. This is annoying an I'm investigating it, but for the time being it isn't the end of the world since a) it still works a lot of the time and b) before, it wasn't even sending any error.
 * This PR should be merged ASAP so that other branches (specifically `feature/s3-config`) can take advantage of the reduced bundle size. Adding `aws-sdk` to the project in that branch made the project un-deployable, again. 
 * While not necessary for production, this PR could significantly benefit from changes I've made to mattdelsordo/claudia-local-api. The development server will crash unless either those changes are merged or you comment out the wildcard/404 endpoint that this PR adds.


## Testing Instructions

 * Start the development server with `./scripts/update` and `./scripts/server`
* You may want to comment out https://github.com/azavea/tilegarden/blob/f5a39163de911a7812216848074642cdbc179a9c/src/tiler/src/api.js#L175-L184 so that the development server doesn't crash (which means you can't test the 404 endpoint in development, unfortunately)
 * Visit [`localhost:3000/`](http://localhost:3000/) and enter some malformed URLs to see what happens


Resolves #75, resolves #76
